### PR TITLE
cmd/snap-update-ns: create missing mount points automatically.

### PR DESCRIPTION
**Pull request by zyga**
This branch allow snap-update-ns to create missing mount points (directories)
automatically.

The code is not perfect yet, we can only create directories as root and we can
only create them with mode 0755. There are ongoing branches that lift those
restrictions but they are not here yet.

The largest restriction is that we cannot yet create directories on top of read
only filesystems. This will be done with a separate feature branch as it is the
most complex.

The apparmor profile for snap-update-ns when invoked from snap-confine was
extended to allow writes to $SNAP_DATA but nowhere else yet. Eventually, to
support the full layout system we will need write permissions in various
places (essentially everywhere except for a specific blacklist).

Signed-off-by: Zygmunt Krynicki <zygmunt.krynicki@canonical.com>

### CODE REVIEW

File: cmd/snap-update-ns/change.go [hunk 1]
```diff
@@ -70,6 +74,29 @@ func (c *Change) Perform() error {
...
+		// was set up by the snap but this is not coded defensively as the code
+		// may execute concurrently with snap application processes.
+		//
+		// To have better defense we may consider freezing the freezer cgroup
+		// belonging to the snap during the execution of snap-update-ns.
+		if _, err := osLstat(c.Entry.Dir); err != nil {
+			if !os.IsNotExist(err) {
+				return err
+			}
+			// TODO: use the right mode and ownership.
```
Comment (zyga) [inline cmd/snap-update-ns/change.go]: This will be addressed after https://github.com/snapcore/snapd/pull/3965 lands.

File: cmd/snap-confine/snap-confine.apparmor.in [hunk 2]
```diff
@@ -492,6 +492,9 @@
         mount options=(rw bind) /var/snap/*/** -> /var/snap/*/**,
         mount options=(ro bind) /var/snap/*/** -> /var/snap/*/**,
 
+        # Allow creating missing mount directories under $SNAP_DATA
+        /var/snap/*/** rw,
```
Comment (jdstrand) [inline cmd/snap-confine/snap-confine.apparmor.in]: To limit to directories, use `/var/snap/*/**/ w,` (do you actually need 'r'?).

File: cmd/snap-update-ns/change.go [hunk 3]
```diff
@@ -70,6 +74,29 @@ func (c *Change) Perform() error {
...
+			if err := osMkdirAll(c.Entry.Dir, 0755); err != nil {
+				return err
+			}
+			if err := osChown(c.Entry.Dir, 0, 0); err != nil {
+				return err
+			}
```
Comment (jdstrand) [inline cmd/snap-update-ns/change.go]: This chown is only doing the bottom directory-- if parents were also created, they would not be chowned. It is probably fine to chown all to root:root for now, but if we decide to allow arbitrary ownership, that will get tricky to limit it to only inside the snap and how to declare which directory is supposed to have which ownership (I'm not sure if that is planned).

File: cmd/snap-update-ns/change.go [hunk 4]
```diff
@@ -70,6 +74,29 @@ func (c *Change) Perform() error {
...
+		// place.  This is very naive, doesn't handle read-only file systems
+		// but it is a good starting point for people working with things like
+		// $SNAP_DATA/subdirectory.
+		//
+		// We use lstat to ensure that we don't follow the symlink in case one
+		// was set up by the snap but this is not coded defensively as the code
+		// may execute concurrently with snap application processes.
+		//
+		// To have better defense we may consider freezing the freezer cgroup
+		// belonging to the snap during the execution of snap-update-ns.
```
Comment (jdstrand) [inline cmd/snap-update-ns/change.go]: I don't like the sound of this. If the snap can race snap-update-ns, then the snap is able to create arbitrary directories anywhere the snap-update-ns profile allows, which is already pretty wide and by your own admission, will only get wider.

We can do better than this. Recall openat/mkdirat in sc_nonfatal_mkpath() in cmd/libsnap-confine-private/utils.c

File: cmd/snap-update-ns/change.go (hunk 3 above)
Comment (jdstrand) [reply cmd/snap-update-ns/change.go]: In thinking about this, I think that root:root 0755 is fine for all of the directories because the snap can always adjust the parent directories after creation. The mountpoint permissions will be covered up by whatever is mounted on it, and the snap could adjust that after the fact as well if desired (but, that will change the permissions of the original mountpoint-- we need to be really care of the mount options for things in SNAP_DATA, SNAP_COMMON or anywhere else the snaps can write to).

File: cmd/snap-confine/snap-confine.apparmor.in (hunk 2 above)
Comment (zyga) [reply cmd/snap-confine/snap-confine.apparmor.in]: We use stat so I assumed we do.

File: cmd/snap-update-ns/change.go (hunk 4 above)
Comment (zyga) [reply cmd/snap-update-ns/change.go]: We can do better wrt mkdir but we cannot ever do better wrt mount itself. I can make the better mkdir for sure but the ultimate race towards mount is unfixable without freezing the snap processes or using some new kernel interface for mounting.

Comment (zyga) [reply cmd/snap-update-ns/change.go]: I've implemented `SecureMkdirAll` and used it place of the original now.

File: cmd/snap-update-ns/change.go (hunk 3 above)
Comment (zyga) [reply cmd/snap-update-ns/change.go]: I'll update the code to ensure we don't leave non root groups (as snap-confine is not setgid root) now.

File: cmd/snap-confine/snap-confine.apparmor.in (hunk 2 above)
Comment (zyga) [reply cmd/snap-confine/snap-confine.apparmor.in]: I'll come up with a few spread tests and if we don't need this I will remove this.

File: cmd/snap-update-ns/change.go (hunk 4 above)
Comment (jdstrand) [reply cmd/snap-update-ns/change.go]: Re mount, yes, we would have to freeze the process since mount() follows symlinks.

Comment (zyga) [reply cmd/snap-update-ns/change.go]: Ack, I'll do that

**Comment by jdstrand** [comment]
I have to say, the timing of this PR is important. I've had two different people walk up to me today at the sprint complaining that they can't use `$SNAP_COMMON/foo` or `$SNAP_DATA/bar` with content interface consuming snaps and I pointed them to this PR. Thanks for taking this on! :)

**Comment by zyga** [comment]
I will fix this branch and add a useful spread test in a moment. Thank you for re-affirming the utility @jdstrand :-)

File: cmd/snap-update-ns/change.go [hunk 5]
```diff
@@ -56,10 +57,116 @@ func (c Change) String() string {
 var (
 	sysMount   = syscall.Mount
 	sysUnmount = syscall.Unmount
+	osLstat    = os.Lstat
+	// NOTE: we're not using os.MkdirAll as it is not careful about symlinks.
+	osMkdirAll = SecureMkdirAll
```
Comment (chipaca) [inline cmd/snap-update-ns/change.go]: why not call the variable `mkdirAll` then?

File: cmd/snap-update-ns/change.go [hunk 6]
```diff
@@ -56,10 +57,116 @@ func (c Change) String() string {
...
+
+// SecureMkdirAll is the secure variant of os.MkdirAll.
+//
+// Unlike the regular version this implementation does not follow any symbolic
+// links. At all times the new directory segment is created using mkdirat(2)
+// while holding an open file descriptor to the parent directory.
+//
+// The only handled error is mkdirat(2) that fails with EEXIST. All other
+// errors are fatal but there is no attempt to undo anything that was created.
+func SecureMkdirAll(name string, perm os.FileMode) error {
```
Comment (chipaca) [inline cmd/snap-update-ns/change.go]: could this live in `osutil` instead?

File: cmd/snap-update-ns/change.go [hunk 7]
```diff
@@ -56,10 +57,116 @@ func (c Change) String() string {
...
+func ensureMountPoint(path string) error {
+	// If the mount point is not present then create a directory in its
+	// place.  This is very naive, doesn't handle read-only file systems
+	// but it is a good starting point for people working with things like
+	// $SNAP_DATA/subdirectory.
+	//
+	// We use lstat to ensure that we don't follow the symlink in case one
+	// was set up by the snap. At the time Change.Perform runs all the
+	// processes in the snap should be frozen.
+	fi, err := osLstat(path)
```
Comment (chipaca) [inline cmd/snap-update-ns/change.go]: why stat it at all, and not dive straight into the mkdirall? as it stands there's a window between the stat and the mkdir that can probably be (ab)used

File: cmd/snap-confine/snap-confine.apparmor.in (hunk 2 above)
Comment (zyga) [reply cmd/snap-confine/snap-confine.apparmor.in]: We need the `r` for the SecureMkdirAll implementation now.

File: cmd/snap-update-ns/change.go (hunk 4 above)
Comment (zyga) [reply cmd/snap-update-ns/change.go]: Done, we are now freezing the group for the update process.

File: cmd/snap-update-ns/change.go [hunk 8]
```diff
@@ -21,6 +21,7 @@ package main
 
 import (
 	"fmt"
+	"os"
 	"path"
```
Comment (chipaca) [inline cmd/snap-update-ns/change.go]: why is this using `path` and not `path/filepath`?

File: cmd/snap-update-ns/change.go (hunk 3 above)
Comment (zyga) [reply cmd/snap-update-ns/change.go]: This is still a TODO

**Comment by zyga** [comment]
@jdstrand I think this ready for another look now. I replied to all the comments and the only remaining TODO is to chown all the path fragments, not just the last one we made. I'll adjust that today but I need to pick one of two ways of doing that (immediately after arch mkdir or after all of them). I'll do a pass on some other PRs that I have and I'll get back here.

File: cmd/snap-update-ns/change.go [hunk 9]
```diff
@@ -70,6 +177,17 @@ func (c *Change) Perform() error {
...
+		if err := ensureMountPoint(c.Entry.Dir); err != nil {
+			return err
+		}
+		// If this is a bind mount then create the source directory as well.
+		// This allows snaps to share a subset of their data easily.
+		if flags&syscall.MS_BIND != 0 {
+			if err := ensureMountPoint(c.Entry.Name); err != nil {
+				return err
+			}
+		}
```
Comment (chipaca) [inline cmd/snap-update-ns/change.go]: checkpoint

File: cmd/snap-update-ns/change.go (hunk 6 above)
Comment (zyga) [reply cmd/snap-update-ns/change.go]: Yes but @jdstrand requested that we limit imports from other modules. I can move it there if he agrees.

File: cmd/snap-update-ns/change.go (hunk 5 above)
Comment (zyga) [reply cmd/snap-update-ns/change.go]: I wanted to keep the same semantics as os.MkdirAll (to make the two interchangeable). I guess it's not a great reason, I can tweak this.

File: cmd/snap-update-ns/change.go (hunk 8 above)
Comment (zyga) [reply cmd/snap-update-ns/change.go]: Hmm, not sure, is there a difference?

File: cmd/snap-update-ns/change.go (hunk 7 above)
Comment (zyga) [reply cmd/snap-update-ns/change.go]: We want to stat it because the error message is different. There's no more any window as the stat is advisory now (mkdirall is coded to defend against that) and all the process are frozen.

File: cmd/snap-update-ns/change.go (hunk 9 above)
Comment (zyga) [reply cmd/snap-update-ns/change.go]: Hmm?

**Comment by zyga** [comment]
@jdstrand I've changed `SecureMkdirAll` to chown each created path segment now. I think this is ready for a security review now please.

File: cmd/snap-update-ns/utils.go [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+		if err != nil {
+			return fmt.Errorf("cannot open root directory, %v", err)
+		}
+	}
+
+	// Split the path by entries and create each element using mkdirat() using
+	// the parent directory as reference. Each time we open the newly created
+	// segment using the O_NOFOLLOW and O_DIRECTORY flag so that symlink
+	// attacks are impossible to carry out.
+	segments := strings.Split(path.Clean(name), "/")
```
Comment (jdstrand) [inline cmd/snap-update-ns/utils.go]: Thank you for using path.Clean()

File: cmd/snap-update-ns/utils.go [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+
+func EnsureMountPointImpl(path string, mode os.FileMode, uid int, gid int) error {
+	// If the mount point is not present then create a directory in its
+	// place.  This is very naive, doesn't handle read-only file systems
+	// but it is a good starting point for people working with things like
+	// $SNAP_DATA/subdirectory.
+	//
+	// We use lstat to ensure that we don't follow the symlink in case one
+	// was set up by the snap. At the time Change.Perform runs all the
+	// processes in the snap should be frozen.
```
Comment (jdstrand) [inline cmd/snap-update-ns/utils.go]: I suggest a stronger statement `// Note that at the time this is run, all the snap's processes are frozen`.

File: cmd/snap-update-ns/freezer_test.go [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+	c.Assert(os.IsNotExist(err), Equals, true)
+
+	// When the freezer cgroup filesystem exists but the particular cgroup
+	// doesn't exist we don nothing at all.
+	c.Assert(os.MkdirAll(update.FreezerCgroupDir(), 0755), IsNil)
+	c.Assert(update.ThawSnapProcesses(n), IsNil)
+	_, err = os.Stat(f)
+	c.Assert(os.IsNotExist(err), Equals, true)
+
+	// When the cgroup exists we write FROZEN the freezer.state file.
```
Comment (jdstrand) [inline cmd/snap-update-ns/freezer_test.go]: This comment is wrong for this test.

File: cmd/snap-update-ns/change.go [hunk 13]
```diff
@@ -53,26 +54,41 @@ func (c Change) String() string {
...
 // Perform executes the desired mount or unmount change using system calls.
 // Filesystems that depend on helper programs or multiple independent calls to
 // the kernel (--make-shared, for example) are unsupported.
 func (c *Change) Perform() error {
-	switch c.Action {
-	case Mount:
-		flags, err := mount.OptsToFlags(c.Entry.Options)
-		if err != nil {
+	if c.Action == Mount {
+		// TODO: use the right mode and ownership.
```
[the remark this answers is not in the document]
Comment (zyga) [reply cmd/snap-update-ns/change.go]: Ack, I'll do that. I suspect this also means we don't need to (urgently) pull in the other PR that adds a way for this to be customized.

File: cmd/snap-update-ns/change.go [hunk 14]
```diff
@@ -53,26 +54,41 @@ func (c Change) String() string {
...
+	}
+	return c.lowLevelPerform()
+}
+
+// lowLevelPerform is a version of Perform that doesn't contain any special behavior.
+func (c *Change) lowLevelPerform() error {
+	switch c.Action {
+	case Mount:
+		flags, unparsed := mount.OptsToCommonFlags(c.Entry.Options)
+		return sysMount(c.Entry.Name, c.Entry.Dir, c.Entry.Type, uintptr(flags), strings.Join(unparsed, ","))
```
[the remark this answers is not in the document]
Comment (zyga) [reply cmd/snap-update-ns/change.go]: Those are needed for things like overlayfs's lower/upper/workdir. I'm not sure if we should have strong validation here given that snap-update-ns reads what snapd wrote. If you feel strongly about it I can add a validation system that will let us be explicit about what is allowed (on top of the mount mediation in apparmor).

File: cmd/snap-update-ns/freezer.go [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"time"
+)
+
+var freezerCgroupDir = "/sys/fs/cgroup/freezer"
+
+func freezeSnapProcesses(snapName string) error {
```
[the remark this answers is not in the document]
Comment (zyga) [reply cmd/snap-update-ns/freezer.go]: I'll add a comment.

The name of the group is `snap.$SNAP` because the attack we are trying to fend off is involving the processes naturally inhabiting the mount namespace and that is no longer per-command but instead per-snap, if you recall.

File: cmd/snap-update-ns/freezer.go [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+		return fmt.Errorf("cannot freeze processes of snap %q, %v", snapName, err)
+	}
+	for i := 0; i < 10; i++ {
+		data, err := ioutil.ReadFile(fname)
+		if err != nil {
+			return fmt.Errorf("cannot determine the freeze state of processes of snap %q, %v", snapName, err)
+		}
+		// If the cgroup is still freezing then wait a moment and try again.
+		if bytes.Equal(data, []byte("FREEZING")) {
+			time.Sleep(100 * time.Millisecond)
```
[the remark this answers is not in the document]
Comment (zyga) [reply cmd/snap-update-ns/freezer.go]: I think this is a multi-socket thing, otherwise the write/freeze would be immediate. I honestly never saw this take any amount of time (kernel expect could perhaps infer this from the inner workings of the cgroup).

File: cmd/snap-update-ns/freezer.go [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+			continue
+		}
+		break
+	}
+	return nil
```
[the remark this answers is not in the document]
Comment (zyga) [reply cmd/snap-update-ns/freezer.go]: Aha, good point! I'll correct that, it was meant to fail here as we check for the failure in the call site.

File: cmd/snap-update-ns/utils.go [hunk 18]
```diff
@@ -0,0 +1,149 @@
...
+//
+// The uid and gid are used for the fchown(2) system call which is performed
+// after each segment is created and opened. The special value -1 may be used
+// to request that ownership is not changed.
+func SecureMkdirAllImpl(name string, perm os.FileMode, uid, gid int) error {
+	// Declare var and don't assign-declare below to ensure we don't swallow
+	// any errors by mistake.
+	var err error
+	// Start at the current working directory by default.
+	var fd int = AT_FDCWD
```
[the remark this answers is not in the document]
Comment (zyga) [reply cmd/snap-update-ns/utils.go]: We don't really need it. I was just trying to write some library-worthy code. I can make it always absolute if you prefer.

File: cmd/snap-update-ns/utils.go [hunk 19]
```diff
@@ -0,0 +1,149 @@
...
+
+	const openFlags = syscall.O_NOFOLLOW | syscall.O_CLOEXEC | syscall.O_DIRECTORY
+
+	// If path is absolute then open the root directory and start there.
+	if path.IsAbs(name) {
+		fd, err = sysOpen("/", openFlags, 0)
```
[the remark this answers is not in the document]
Comment (zyga) [reply cmd/snap-update-ns/utils.go]: Yeah, that's not something we can avoid in the go binding to the syscall.

File: cmd/snap-update-ns/freezer.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (zyga) [reply cmd/snap-update-ns/freezer.go]: Man, I just realized I misread your comment :)

File: cmd/snap-update-ns/freezer.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (zyga) [reply cmd/snap-update-ns/freezer.go]: For the sake of being conservative I increased this to three seconds.

File: cmd/snap-update-ns/freezer.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (zyga) [reply cmd/snap-update-ns/freezer.go]: Corrected. I also made this thaw the processes so that we don't break the system when this happens.

File: cmd/snap-update-ns/freezer_test.go (hunk 12 above)
Comment (zyga) [reply cmd/snap-update-ns/freezer_test.go]: Thank you, corrected.

File: cmd/snap-update-ns/utils.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (zyga) [reply cmd/snap-update-ns/utils.go]: Done.

File: cmd/snap-update-ns/change.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (jdstrand) [reply cmd/snap-update-ns/change.go]: Not necessarily for this PR, but I feel it is worthwhile to have input validation for maintenance' sake. A typo or something else is caught in unit tests as opposed to spread tests.

**Comment by zyga** [comment]
@jdstrand all of the comments have been addressed! Thank you for the review :)

File: cmd/snap-update-ns/utils.go [hunk 20]
```diff
@@ -0,0 +1,144 @@
...
+// to request that ownership is not changed.
+func SecureMkdirAllImpl(name string, perm os.FileMode, uid, gid int) error {
+	// Declare var and don't assign-declare below to ensure we don't swallow
+	// any errors by mistake.
+	var err error
+	var fd int
+
+	const openFlags = syscall.O_NOFOLLOW | syscall.O_CLOEXEC | syscall.O_DIRECTORY
+
+	if !path.IsAbs(name) {
```
Comment (jdstrand) [inline cmd/snap-update-ns/utils.go]: If you are going to make another commit, a comment here would be nice saying that we actively chose not to support relative directories. Eg:
```
// Only support absolute paths to avoid bugs in snap-confine when
// called from anywhere
```

Comment (zyga) [reply cmd/snap-update-ns/utils.go]: Sure, I'll add this before merging. Thank you!

File: cmd/snap-update-ns/utils.go [hunk 21]
```diff
@@ -0,0 +1,144 @@
...
+				return fmt.Errorf("cannot mkdir path segment %q, %v", segment, err)
+			}
+		}
+		previousFd := fd
+
+		fd, err = sysOpenat(fd, segment, openFlags, 0)
+		if err := sysClose(previousFd); err != nil {
+			return fmt.Errorf("cannot close previous file descriptor, %v", err)
+		}
+		if err != nil {
```
Comment (mvo5) [inline cmd/snap-update-ns/utils.go]: Why do we have the `sysClose()` between this error check? The error is for the err in 101 so it feels harder to read this way than if the err check is followed immediately after the sysOpenat(). Maybe defer sysClose() would help to make this easier to read?

File: cmd/snap-update-ns/utils.go [hunk 22]
```diff
@@ -0,0 +1,144 @@
...
+		}
+
+	}
+	if err = sysClose(fd); err != nil {
+		return fmt.Errorf("cannot close file descriptor, %v", err)
+	}
+	return nil
+}
+
+func EnsureMountPointImpl(path string, mode os.FileMode, uid int, gid int) error {
```
Comment (mvo5) [inline cmd/snap-update-ns/utils.go]: I think this can be renamed to "EnsureMountPoint()" (or possible even ensureMountPoint) and the `var ensureMountPoint` can be removed. AFAICT this is not reassigned anywhere in the code.

File: tests/lib/snaps/test-snapd-content-advanced-plug/meta/snap.yaml [hunk 23]
```diff
@@ -0,0 +1,13 @@
+name: test-snapd-content-advanced-plug
+summary: A more complex snap for testing content interface
```
Comment (mvo5) [inline tests/lib/snaps/test-snapd-content-advanced-plug/meta/snap.yaml]: Thanks for this test!

File: cmd/snap-update-ns/main.go [hunk 24]
```diff
@@ -101,6 +101,14 @@ func run() error {
 		}
 	}
 
+	// Freeze the mount namespace and unfreeze it later. This lets us perform
+	// modifications without snap processes attempting to construct symlinks or
+	// perform other malicious activity.
```
[the remark this answers is not in the document]
Comment (zyga) [reply cmd/snap-update-ns/main.go]: It means attempting to mount a symlink attack against mount. I'll adjust the comment to be explicit about this.

File: cmd/snap-update-ns/change.go [hunk 25]
```diff
@@ -53,26 +54,40 @@ func (c Change) String() string {
...
+		if flags&syscall.MS_BIND != 0 {
+			if err := ensureMountPoint(c.Entry.Name, mode, uid, gid); err != nil {
+				return err
+			}
+		}
+	}
+	return c.lowLevelPerform()
+}
+
+// lowLevelPerform is a version of Perform that doesn't contain any special behavior.
```
[the remark this answers is not in the document]
Comment (zyga) [reply cmd/snap-update-ns/change.go]: They are separated because code further ahead needs to call lowLevelPerform. I'll adjust comments on this but I need them separate in one way or another.

File: cmd/snap-update-ns/utils.go (hunk 22 above)
Comment (zyga) [reply cmd/snap-update-ns/utils.go]: +1

File: tests/lib/snaps/test-snapd-content-advanced-plug/meta/snap.yaml (hunk 23 above)
Comment (zyga) [reply tests/lib/snaps/test-snapd-content-advanced-plug/meta/snap.yaml]: More is coming here :) This is just a start

File: cmd/snap-update-ns/utils.go (hunk 21 above)
Comment (zyga) [reply cmd/snap-update-ns/utils.go]: The code is designed to ensure that we only keep two file descriptors around. The one for the parent directory and the one for the entry inside it. Everything is paired correctly.

As for defer, we could do that but it would mean we need to keep more file descriptors open for no good reason.

File: cmd/snap-update-ns/utils.go [hunk 26]
```diff
@@ -0,0 +1,144 @@
...
+			switch err {
+			case syscall.EEXIST:
+				made = false
+			default:
+				return fmt.Errorf("cannot mkdir path segment %q, %v", segment, err)
+			}
+		}
+		previousFd := fd
+
+		fd, err = sysOpenat(fd, segment, openFlags, 0)
```
[the remark this answers is not in the document]
Comment (zyga) [reply cmd/snap-update-ns/utils.go]: See below.

File: cmd/snap-update-ns/utils.go [hunk 27]
```diff
@@ -0,0 +1,144 @@
...
+	const openFlags = syscall.O_NOFOLLOW | syscall.O_CLOEXEC | syscall.O_DIRECTORY
+
+	if !path.IsAbs(name) {
+		return fmt.Errorf("cannot create directory with relative path: %q", name)
+	}
+	// Open the root directory and start there.
+	fd, err = sysOpen("/", openFlags, 0)
+	if err != nil {
+		return fmt.Errorf("cannot open root directory, %v", err)
+	}
```
[the remark this answers is not in the document]
Comment (zyga) [reply cmd/snap-update-ns/utils.go]: Indeed, let me rework this.

File: cmd/snap-update-ns/utils.go (hunk 20 above)
Comment (zyga) [reply cmd/snap-update-ns/utils.go]: Done

File: cmd/snap-update-ns/main.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (zyga) [reply cmd/snap-update-ns/main.go]: Done now

File: cmd/snap-update-ns/utils.go (hunk 22 above)
Comment (zyga) [reply cmd/snap-update-ns/utils.go]: Done now

File: cmd/snap-update-ns/change.go (hunk 25 above)
[the remark this answers is not in the document]
Comment (zyga) [reply cmd/snap-update-ns/change.go]: Done now

File: cmd/snap-update-ns/utils.go [hunk 28]
```diff
@@ -0,0 +1,137 @@
...
+// The uid and gid are used for the fchown(2) system call which is performed
+// after each segment is created and opened. The special value -1 may be used
+// to request that ownership is not changed.
+func secureMkdirAll(name string, perm os.FileMode, uid, gid int) error {
+	// Declare var and don't assign-declare below to ensure we don't swallow
+	// any errors by mistake.
+	var err error
+	var fd int
+
+	const openFlags = syscall.O_NOFOLLOW | syscall.O_CLOEXEC | syscall.O_DIRECTORY
```
Comment (chipaca) [inline cmd/snap-update-ns/utils.go]: I'd suggest also `O_PATH`. But you need to use `golang.org/x/sys/unix` instead of `syscall` (which we should do as much as possible anyway)

File: cmd/snap-update-ns/utils.go [hunk 29]
```diff
@@ -0,0 +1,135 @@
...
+				made = false
+			default:
+				return fmt.Errorf("cannot mkdir path segment %q: %v", segment, err)
+			}
+		}
+		fd, err = sysOpenat(fd, segment, openFlags, 0)
+		if err != nil {
+			return fmt.Errorf("cannot open path segment %q: %v", segment, err)
+		}
+		defer sysClose(fd)
```
[the remark this answers is not in the document]
Comment (zyga) [reply cmd/snap-update-ns/utils.go]: Look at the history, one of the most recent patches introduces defer, if we pop that we're back to hand-done cleanup logic that only uses a pair of FDs

File: cmd/snap-update-ns/utils.go [hunk 30]
```diff
@@ -0,0 +1,135 @@
...
+		if made {
+			// Chown each segment that we made.
+			if err := sysFchown(fd, uid, gid); err != nil {
+				return fmt.Errorf("cannot chown path segment %q to %d.%d: %v", segment, uid, gid, err)
+			}
+		}
+
+	}
+	return nil
```
[the remark this answers is not in the document]
Comment (zyga) [reply cmd/snap-update-ns/utils.go]: No, this is not a requirement of this operation.

File: cmd/snap-update-ns/utils.go [hunk 31]
```diff
@@ -0,0 +1,135 @@
...
+	if err != nil {
+		return fmt.Errorf("cannot open root directory: %v", err)
+	}
+	defer sysClose(fd)
+
+	// Split the path by entries and create each element using mkdirat() using
+	// the parent directory as reference. Each time we open the newly created
+	// segment using the O_NOFOLLOW and O_DIRECTORY flag so that symlink
+	// attacks are impossible to carry out.
+	segments := strings.Split(filepath.Clean(name), "/")
```
[the remark this answers is not in the document]
Comment (zyga) [reply cmd/snap-update-ns/utils.go]: Done

File: cmd/snap-update-ns/utils.go (hunk 28 above)
Comment (zyga) [reply cmd/snap-update-ns/utils.go]: Done

Comment (zyga) [reply cmd/snap-update-ns/utils.go]: Undone because O_PATH cannot be used with fchown.
